### PR TITLE
add collections_04 task

### PR DESCRIPTION
**Pull request by wilden5**
add store commands

### CODE REVIEW

File: consoleApp/src/main/java/com/denis/consoleapp/StoreApp.java [hunk 1]
```diff
@@ -2,10 +2,49 @@
...
+    }
+
+    private static void executeStore(Scanner scanner, Store store) {
+        System.out.println("Available list of commands: print, sort, top, quit");
+        String storeCommand = scanner.nextLine();
+        switch (storeCommand) {
+            case print:
+                System.out.println("Store returned to standard state"); // sort and top commands not modifying product list
+                store.print();
+                executeStore(scanner, store);
```
Comment (artemmoroz) [inline consoleApp/src/main/java/com/denis/consoleapp/StoreApp.java]: * try to remove it from here

File: store/src/main/java/com/denis/store/Store.java [hunk 2]
```diff
@@ -30,4 +36,55 @@ public String toString() {
     public void print() {
         System.out.println(this);
     }
+
+    public void printSort() {
```
Comment (artemmoroz) [inline store/src/main/java/com/denis/store/Store.java]: SOLID, move somewhere

Comment (artemmoroz) [reply store/src/main/java/com/denis/store/Store.java]: for al print methods

File: store/src/main/java/com/denis/store/Store.java [hunk 3]
```diff
@@ -30,4 +36,55 @@ public String toString() {
...
+        CommandSortComparator commandSortComparator = new CommandSortComparator();
+        List<Category> cloneCategory = new ArrayList<>();
```
Comment (artemmoroz) [inline store/src/main/java/com/denis/store/Store.java]: line 41-55 -> Clonable inteface (deep clone)

File: store/src/main/java/com/denis/store/Store.java [hunk 4]
```diff
@@ -30,4 +36,55 @@ public String toString() {
...
+    private List<Product> getTopSortProduct() {
+        ArrayList<Product> allProducts = new ArrayList<>();
+
+        Comparator priceComparator = new CommandTopComparator();
+        for (Category cat : categoryList) {
+            allProducts.addAll(cat.getProductList());
+        }
+
+        Collections.sort(allProducts, priceComparator);
+        int indexCut = (allProducts.size() > 5) ? 5 : allProducts.size();
```
Comment (artemmoroz) [inline store/src/main/java/com/denis/store/Store.java]: list.stream.limit

File: store/src/main/java/com/denis/store/utility/CommandSortComparator.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package com.denis.store.utility;
+
+import com.denis.domain.Product;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.function.Function;
+
+public class CommandSortComparator implements Comparator<Product> {
+    List<XmlReader.SortRules> sortRules;
```
Comment (artemmoroz) [inline store/src/main/java/com/denis/store/utility/CommandSortComparator.java]: single class (map)

File: store/src/main/java/com/denis/store/utility/XmlReader.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import javax.xml.stream.XMLStreamException;
+import javax.xml.stream.XMLStreamReader;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+public class XmlReader {
+    public static List<SortRules> parceXml() {
+        List<SortRules> fieldToOrder = new ArrayList<>();
+        InputStream istream = XmlReader.class.getResourceAsStream("SortParams.xml");
```
Comment (artemmoroz) [inline store/src/main/java/com/denis/store/utility/XmlReader.java]: filename - move to func params

**Review by artemmoroz — CHANGES_REQUESTED** [review]
solid, flexible functions

**Comment by wilden5** [comment]
solved the noted issues

File: store/src/main/java/com/denis/store/utility/CommandSortComparator.java (hunk 5 above)
Comment (artemmoroz) [reply store/src/main/java/com/denis/store/utility/CommandSortComparator.java]: what about that comment? as i remember, we decided to make generic comparator, that will use some generic settings for comparing, like map<string, string> and not bound to xml reader or so.

**Review by artemmoroz — APPROVED** [review]

**Pull request merged**